### PR TITLE
[SPARK-6650] [core] Stop ExecutorAllocationManager when context stops.

### PR DESCRIPTION
**Pull request by vanzin**
This fixes the thread leak. I also changed the unit test to keep track
of allocated contexts and make sure they're closed after tests are
run; this is needed since some tests use this pattern:

```
val sc = createContext()
doSomethingThatMayThrow()
sc.stop()
```

### CODE REVIEW

File: core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala [hunk 1]
```diff
@@ -28,10 +28,20 @@ import org.apache.spark.util.ManualClock
 /**
  * Test add and remove behavior of ExecutorAllocationManager.
  */
-class ExecutorAllocationManagerSuite extends FunSuite with LocalSparkContext {
+class ExecutorAllocationManagerSuite extends FunSuite with LocalSparkContext with BeforeAndAfter {
   import ExecutorAllocationManager._
   import ExecutorAllocationManagerSuite._
 
+  private val contexts = new mutable.ListBuffer[SparkContext]()
```
Comment (sryza) [inline core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala]: Will this ever have more than a single element?

Comment (vanzin) [inline core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala]: Yes. See ` test("verify min/max executors")`.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala [hunk 2]
```diff
@@ -173,32 +179,24 @@ private[spark] class ExecutorAllocationManager(
...
-          Thread.sleep(intervalMillis)
-        }
-      }
-    }
-    t.setName("spark-dynamic-executor-allocation")
-    t.setDaemon(true)
-    t.start()
+  def stop(): Unit = {
+    executor.shutdown()
+    executor.awaitTermination(10, TimeUnit.SECONDS)
```
Comment (sryza) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: Thinking aloud here:
We stop the ExecutorAllocationManager after stopping the DAGScheduler, which means, in yarn-client mode, we'll have already torn down the YARN application.  If `schedule` is called after that, we could be trying to make an RPC to the AM, which is no longer there.  So it seems like waiting the full 10 seconds for that RPC to time out could be a common case.

Comment (vanzin) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: I tried to follow the order of initialization during shutdown (the alloc manager is started after the event logger so should be stopped right before it - yeah, I know my code doesn't do exactly that).

But we could move the stop before the scheduler's stop.

**Comment by vanzin** [comment]
Jenkins retest this please.

**Comment by srowen** [comment]
I'm not sure why the test failed, but on inspection it LGTM.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala [hunk 3]
```diff
@@ -17,10 +17,12 @@
 
 package org.apache.spark
 
+import java.util.concurrent.{TimeUnit, Executors}
```
Comment (sryza) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: nit: alphabetize

**Comment by sryza** [comment]
One last nit.  Otherwise LGTM.

**Comment by andrewor14** [comment]
LGTM2, thanks for fixing this merging into master and 1.3.

**Pull request closed without merging**